feat(sesv2): Implement Simple Email Service v2 resources

kelvinwijaya opened the pull request:
<!--
Thank you for helping to improve Crossplane!

Please read through https://git.io/fj2m9 if this is your first time opening a
Crossplane pull request. Find us in https://slack.crossplane.io/messages/dev if
you need any help contributing.
-->

### Description of your changes

<!--
Briefly describe what this pull request does. Be sure to direct your reviewers'
attention to anything that needs special consideration.

We love pull requests that resolve an open Crossplane issue. If yours does, you
can uncomment the below line to indicate which issue your PR fixes, for example
"Fixes #500":

-->
Fixes #414 

I have:

- [X] Read and followed Crossplane's [contribution process].
- [X] Run `make reviewable test` to ensure this PR is ready for review.

### How has this code been tested
- Tried Create resource operation with examples of custom resource manifests
- Make sure Readiness Status shown as `True` and Condition as `Available` when the resources are provisioned successfully
- Monitor Readiness Status shown as `False` and Condition as `Unavailable` OR `Creating` if wrong parameters being supplied in the custom resource manifests
- Tried various Update resource operations depending on the targetted functionality especially when there is more than 1 PUT API operations.  
- Tried Delete operation of the resource and get deleted successfully

<!--
Before reviewers can be confident in the correctness of this pull request, it
needs to tested and shown to be correct. Briefly describe the testing that has
already been done or which is planned for this change.
-->

Test plan based on examples template:
1. Test `emailidentity` CRD operations for email address (verified with email acknowledgement status)
2. Test `emailidentity` CRD operations for domain (verified with DKIM and MailFrom status)
3. Test `emailidentity` Update operations for `ConfigurationSetAttributes`
4. Test `emailidentity` Update operations for `MailFromAttributes`
5. Test `configurationset` CRD operations which is referenced in the emailidentity
6. Test `configurationset` Update operations for `DeliveryOptions`
7. Test `configurationset` Update operations for `ReputationOptions`
8. Test `configurationset` Update operations for `SuppressionOptions`
9. Test `configurationset` Update operations for `SendingOptions`
10. Test `configurationset` Update operations for `TrackingOptions`
11. Test `emailtemplate` CRUD operations with `HTML,Subject,Text`

[contribution process]: https://git.io/fj2m9

kelvinwijaya commented:
Hi @muvaf @negz 

I have just submitted a PR, not sure what causing the CI/check-diff step failing, can advise on this?

Thanks,
Kelvin

chlunde commented:
@kelvinwijaya yeah, I think it's either
* because you modified a generated file
* or didn't run `make generate` after rebasing/with the same tools, or after a local change

try `make check-diff` locally (make generate; git diff)

chlunde commented on apis/sesv2/v1alpha1/zz_email_i_dentity.go:
```diff
@@ -0,0 +1,113 @@
+/*
```
I think this should be `identity`, not `i_dentity` in `zz_email_i_dentity.go`. I wonder if this has been fixed upstream in the latest version of https://github.com/aws-controllers-k8s/code-generator/ - maybe we should bump that as a PR first, or create an issue there.

chlunde replied:
This has been fixed upstream in code-generator

kelvinwijaya replied:
Does it mean I have to use different commit for `CODE_GENERATOR` to get this item fixed?

chlunde replied:
yes, and there's a PR for that already: https://github.com/crossplane/provider-aws/pull/801

chlunde commented on apis/sesv2/v1alpha1/custom_types.go:
```diff
@@ -0,0 +1,58 @@
+type CustomEmailTemplateParameters struct{}
+
+// CustomEmailIDentityObservation are parameters for
+type CustomEmailIDentityObservation struct {
```
With `SERVICES=sesv2 CODE_GENERATOR_COMMIT=main make services` this becomes `Identity` instead of `IDentity`, so I guess the file name issue should be reported..

chlunde commented on apis/sesv2/v1alpha1/generator-config.yaml:
```diff
@@ -0,0 +1,9 @@
+ignore:
```
ImportJobSummary, BlacklistEntry, Contact, DailyVolume, CustomVerificationEmailTemplateMetadata - I notice a few things hand-edited in the generated files, could you add them here?

kelvinwijaya replied:
Thanks @chlunde for highlighting, Those structs are not hand-edited by me, I notice that is it auto-generated probably because i didn't add the mentioned list in the `ignore:` items. Once we fixed the `check-diff` issue below, I will push all the changes again.

kelvinwijaya replied:
Hi @chlunde , I have added the list of `resources` and `field_paths` to resolve unused one

kelvinwijaya commented:
> @kelvinwijaya yeah, I think it's either
> 
> * because you modified a generated file
> * or didn't run `make generate` after rebasing/with the same tools, or after a local change
> 
> try `make check-diff` locally (make generate; git diff)

Thanks @chlunde 

I have rebased my code and removed handwritten code from the auto-generated one. But i noticed errors during the go building process:
`ERRO Running error: buildir: analysis skipped: errors in package: [/Users/kelvinwijaya/go/src/github.com/kelvinwijaya/provider-aws/pkg/controller/sesv2/configurationseteventdestination/zz_conversions.go:48:9: elem declared but not used] 
09:06:52 [FAIL]`

Problem is if I don't modify this file to make it work as below, it wont compile:
Original:
```
for _, elem := range resp.EventDestinations {
		found = true
		break
	}
```
After modification (no compile error):
```
for _, elem := range resp.EventDestinations {
		if elem.Name != nil {
			found = true
			break
		}
	}
```

But after making this modification and commit the code. I ran `make generate` again and issue is this `zz_conversion.go` file will be overwritten again by the code-generator as there is diff. How do we resolve this before if i push this code? Else,  the `CI/check-diff` will definitely throw error again.

AaronME commented:
/size L

muvaf commented:
@kelvinwijaya I wonder why the content of the for loop is empty. From your message in Slack, it seems it has some fields that should be used in the for loop. I think we need to open an issue there and to get the PR going, you can add this specific resource to the ignore list. Once the issue is addressed there, we can get back to adding it.

kelvinwijaya commented:
Thanks @muvaf for your input, i will open a new case in `aws-controller-k8s/code-generator` project. I have excluded the problematic resource `configurationseteventdestination` so that this PR can keep going. This resource is not essential for SES to work , so I think I am good as for now and will definitely add it back when the code-generator fixes has been made

Thanks @chlunde also for pointing out the issue in code-generator, I will follow-up by opening an issue ticket with them

I have make sure `All check have passed` and would require you and other maintainers for the reviews. Many thanks in advanced.

kelvinwijaya commented:
I have open a case to `aws-controllers-k8s/community` on the code-generation issue: https://github.com/aws-controllers-k8s/community/issues/898

kelvinwijaya commented:
Hi @chlunde 

I have added Update operations for TrackingOptions. Feel free to review my PR whenever you ready or available

AaronME commented:
@chlunde Would you mind taking another look and confirming your comments are resolved?

chlunde commented on examples/sesv2/emailidentity-domain.yaml:
```diff
@@ -0,0 +1,23 @@
+    mailFromAttributes:
+      behaviorOnMxFailure: USE_DEFAULT_VALUE #USE_DEFAULT_VALUE | REJECT_MESSAGE
+      mailFromDomain: subdomain.mydomain
+    emailIDentity: mydomain
```
@AaronME I think this (`emailIDentity` -> `emailIdentity`) will be fixed with the new code gen version. So the question is if we should wait for that first, or just make sure we get both merged before we do a release. Either way, one of them will have to be rebased. I think I lean on getting the SDK/code gen PRs merged first. They are fairly large and harder to review, and if we do the new resource PRs first we might end up with more complex "re-reviews" and rebases on the SDK PRs.

As for the rest of the code, I have not reviewed it fully, I just looked over it to help @kelvinwijaya forward due to lack of reviews.

I think it's also a question of what quality and review we should have for *alpha* resources. It might be good to get them merged, and then be stricter for alpha->beta graduation, when we have someone who knows the product and can follow up with fixes in the future, and we don't have enough maintainer capacity to fully understand and test the specific service.

kelvinwijaya commented:
Hi @chlunde @AaronME 

Would like to check if this PR will be pushed to the next `provider-aws` release

kelvinwijaya commented:
Adding comment for reminder as this PR is currently blocked by:
#801 has been closed and replaced by #918

kelvinwijaya commented:
Hi @chlunde @AaronME

It has been a while that this PR is blocked due to #801 #918, Is there a way we can push this implementation first while waiting for that. It was quite troublesome for my organization as we have to maintain downstream repo to support this additional feature

kelvinwijaya commented:
> 

Hi @chlunde @AaronME

Please ignore my comment earlier. 

I notice the **code-generator** is now referencing to newer version (v0.14.1) with commit: 285d87b66b62fbfb859986ddf74c9f9b6ae743fb

Let me update the changes accordingly and update your guys with the new commit, I believe this should fix the `zz_email_i_dentity.go` and `CustomEmailIDentityObservation` and hopefully we can close this issue.


Thanks

kelvinwijaya commented:
Hi @chlunde

I have done with my changes using the latest code-generator
Please help me review this PR again.

Many thanks!

kelvinwijaya commented:
Hi @haarchri

Saw the upcoming release for provider-aws for v0.24.0: https://github.com/crossplane/provider-aws/issues/1061

Currently this PR is still pending for review, will there be anyone in community to help out on this? I really hope this SES resource can be integrated to the aws provider release soon.

muvaf commented on apis/sesv2/v1alpha1/zz_email_template.go:
```diff
@@ -0,0 +1,92 @@
+	TemplateContent *EmailTemplateContent `json:"templateContent"`
+	// The name of the template you want to create.
+	// +kubebuilder:validation:Required
+	TemplateName                  *string `json:"templateName"`
```
This should be taken from [external name annotation](https://github.com/crossplane/provider-aws/blob/master/CODE_GENERATION.md#external-name) I think.

kelvinwijaya replied:
i have changed using external-name annotation referencing

muvaf commented on apis/sesv2/v1alpha1/zz_email_identity.go:
```diff
@@ -0,0 +1,110 @@
+	// The configuration set to use by default when sending from this identity.
+	// Note that any configuration set defined in the email sending request takes
+	// precedence.
+	ConfigurationSetName *string `json:"configurationSetName,omitempty"`
```
I think we can have a [referencer](https://github.com/crossplane/provider-aws/blob/master/CODE_GENERATION.md#referencer-fields) here for `ConfigurationSet` type.

muvaf commented on pkg/controller/sesv2/configurationset/setup.go:
```diff
@@ -0,0 +1,288 @@
+
+// isUpToDateReputationOptions checks if ReputationOptions Object are up to date
+func isUpToDateReputationOptions(cr *svcapitypes.ConfigurationSet, resp *svcsdk.GetConfigurationSetOutput) bool {
+	if cr.Spec.ForProvider.ReputationOptions != nil && resp.ReputationOptions != nil {
```
Is there a case where the `resp.ReputationOptions` is `nil` because user didn't give it in the beginning and then after they update the spec, we need to issue an update in that situation? i.e. should we call in case where `cr.Spec.ForProvider.ReputationOptions != nil && resp.ReputationOptions == nil` is true? Or is `resp.ReputationOptions` always returned with some defaults?

muvaf replied:
Goes for other up-to-date comparisons as well.

kelvinwijaya replied:
Check against the AWS API calls
Following options will be returned regardless:
- DeliveryOptions
- ReputationOptions
- SendingOptions

Following options depends on user toggle:
- TrackingOptions
- SuppressionOptions

Will rectify based on these 2 options

kelvinwijaya replied:
Anyway, i have rectified these in my previous commits

muvaf commented on pkg/controller/sesv2/configurationset/setup.go:
```diff
@@ -0,0 +1,288 @@
+		return aws.StringValue(i) < aws.StringValue(j)
+	})
+
+	return cmp.Equal(crSuppressedReasons, awsSuppressedReasons, sortCmp)
```
I think a simple for loop checking for `StringValue` would be less costly here than reflection.

kelvinwijaya replied:
for loop method will be more verbose (less clarity) and input from custom resource may not be in order

muvaf commented on pkg/controller/sesv2/configurationset/setup.go:
```diff
@@ -0,0 +1,288 @@
+	// Update Resource is not provided other than individual PUT operation
+	// NOTE: Update operation NOT generated by ACK code-generator
+
+	// update DeliveryOptions (PutConfigurationSetDeliveryOptions)
```
Would it be possible for us to issue a `GET` call here and see what we need to update exactly so that we make less calls?

kelvinwijaya replied:
Can you show me the concrete example on how i can achieve this?

Is there a way we can achieve via passing info from `IsUpToDate` as comparison of both cr and actual resource state is happening here?

MisterMX replied:
You can add fields to the `hooks` struct, set them during `Observe` and access them again during `Update` or `Create`.

kelvinwijaya replied:
Will have a look on that shortly. If i understand correctly, there is only `preObserve` and `postObserve` func that can be overridden in the controller and for this case, `postObserve` would be a fit

kelvinwijaya replied:
Have enhanced the update function to trigger only based on specific attribute changes

muvaf commented on pkg/controller/sesv2/tags.go:
```diff
@@ -0,0 +1,154 @@
+}
+
+// UpdateTagsForResource with resourceName
+func UpdateTagsForResource(client sesv2iface.SESV2API, spec []*svcapitypes.Tag, resourceArn *string) error {
```
Seems like this is never used.

kelvinwijaya replied:
Yes, keeping the code once able to figure out: // TODO(kelvinwijaya): Updating Tags will be useful if we can determine the resourceArn

MisterMX replied:
Whats blocking this TODO from being implemented?

kelvinwijaya replied:
As for ConfigurationSet resource, the update of Tags is passed via CreateConfigurationSet API only: https://docs.aws.amazon.com/ses/latest/APIReference-V2/API_CreateConfigurationSet.html

As for today, the tagging API doesn't support sub resources tagging (e.g. configuratinset) and the only resource that can use this is emailidentity which has the arn based resource name: https://docs.aws.amazon.com/ses/latest/APIReference-V2/API_TagResource.html

I am keeping the code for future usage, but i can remove if this cause confusion as not being used for now.

muvaf commented on pkg/controller/sesv2/tags.go:
```diff
@@ -0,0 +1,154 @@
+
+	tags := spec
+	for _, t := range GetExternalTags(mg) {
+		if _, exists := tagMap[awsclient.StringValue(t.Key)]; !exists {
```
We should override its value if it exists.

MisterMX replied:
I am not sure on this one. If the user specifies a different value in the spec the controller should keep it. We some issues with this in the past.

I would keep it the way it is and only set tag values if the key isn't present already.

muvaf commented on pkg/controller/sesv2/tags.go:
```diff
@@ -0,0 +1,154 @@
+}
+
+// AddExternalTags to spec if they don't exist
+func AddExternalTags(mg resource.Managed, spec []*svcapitypes.Tag) []*svcapitypes.Tag {
```
Would it be possible to have an implementation similar to this one that covers all cases including composition and ordering? https://github.com/crossplane/provider-aws/blob/8be1568/pkg/controller/ecr/repository/controller.go#L246

muvaf reviewed (commented):
Thanks @kelvinwijaya !

haarchri commented:
@kelvinwijaya any chance to check the Open Points that we can finalize this PR ?

kelvinwijaya commented:
Hi @haarchri 

Let me have a look at the open points and try to resolve the review comments

MisterMX commented:
@kelvinwijaya can you fix the issues and rebase to the latest master? Then we can work on getting this merged.

kelvinwijaya commented:
@MisterMX , absolutely.. let me plan on review issues and merge the changes

kelvinwijaya commented on apis/sesv2/v1alpha1/generator-config.yaml:
```diff
@@ -0,0 +1,25 @@
+ignore:
+  field_paths:
```
Yes. those fields are not from the 3 resources except the one with comments which fail during generation. Those resources that are excluded is not the core SES offerings like status reporting and templating and some requires fixes from ack code-generator

kelvinwijaya commented on pkg/controller/sesv2/emailidentity/setup.go:
```diff
@@ -0,0 +1,204 @@
+			},
+			SigningAttributesOrigin: awsclient.String(dkimSigningAttributeExternal),
+		}
+		if _, err := e.client.PutEmailIdentityDkimSigningAttributesWithContext(ctx, dkimSigningAttributesInput); err != nil {
```
Agree to move to `Update` function to cater for the design

kelvinwijaya commented on pkg/controller/sesv2/emailidentity/setup.go:
```diff
@@ -0,0 +1,204 @@
+		dkimSigningAttributesInput = &svcsdk.PutEmailIdentityDkimSigningAttributesInput{
+			EmailIdentity: cr.Spec.ForProvider.EmailIdentity,
+			SigningAttributes: &svcsdk.DkimSigningAttributes{
+				DomainSigningPrivateKey: cr.Spec.ForProvider.DkimSigningAttributes.DomainSigningPrivateKey,
```
Enable the option to store dkim private key via Secret Object Reference

kelvinwijaya commented:
@MisterMX , for your review https://github.com/crossplane-contrib/provider-aws/pull/791/commits/8e9104b93708a9632c174aa2344daa4399c2be06

kelvinwijaya commented on apis/sesv2/v1alpha1/zz_configuration_set.go:
```diff
@@ -0,0 +1,105 @@
+	Region string `json:"region"`
+	// The name of the configuration set.
+	// +kubebuilder:validation:Required
+	ConfigurationSetName *string `json:"configurationSetName"`
```
i have changed using external-name annotation referencing

kelvinwijaya commented on apis/sesv2/v1alpha1/zz_email_identity.go:
```diff
@@ -0,0 +1,110 @@
+	DkimSigningAttributes *DkimSigningAttributes `json:"dkimSigningAttributes,omitempty"`
+	// The email address or domain that you want to verify.
+	// +kubebuilder:validation:Required
+	EmailIdentity *string `json:"emailIdentity"`
```
Not changing to external name annotation in consideration with email identity contain characters (e.g. my@acme.com) that is not valid k8s metadata name

kelvinwijaya commented:
@MisterMX, added  changes based on some reviews that i missed as conversation is not auto-populated in `File changed` due to comment given for autogenerated files (zz_files): https://github.com/crossplane-contrib/provider-aws/pull/791/commits/b9cd5da9c42ad10251b0477b815316f3e2bdcae9

MisterMX commented on pkg/controller/sesv2/configurationset/setup.go:
```diff
@@ -0,0 +1,345 @@
+func isUpToDateTrackingOptions(cr *svcapitypes.ConfigurationSet, resp *svcsdk.GetConfigurationSetOutput) bool {
+	// User may remove TrackingOptions spec and expect an update
+	if cr.Spec.ForProvider.TrackingOptions == nil && resp.TrackingOptions != nil {
+		cr.Spec.ForProvider.TrackingOptions = &svcapitypes.TrackingOptions{}
```
`cr` should not be changed during `isUpTodate`.

kelvinwijaya replied:
have fixed this issue

MisterMX commented on pkg/controller/sesv2/configurationset/setup.go:
```diff
@@ -0,0 +1,345 @@
+	// User may remove SuppressionOptions from spec and expect an update
+	if cr.Spec.ForProvider.SuppressionOptions == nil && resp.SuppressionOptions != nil {
+		// PutConfigurationSetSuppressionOptions requires empty slice to toggle update
+		cr.Spec.ForProvider.SuppressionOptions = &svcapitypes.SuppressionOptions{SuppressedReasons: []*string{}}
```
`cr` should not be changed during `isUpTodate`.

kelvinwijaya replied:
have fixed this issue

MisterMX commented on pkg/controller/sesv2/configurationset/setup.go:
```diff
@@ -0,0 +1,345 @@
+	awsSuppressedReasons := make([]*string, 0)
+
+	if cr.Spec.ForProvider.SuppressionOptions != nil && cr.Spec.ForProvider.SuppressionOptions.SuppressedReasons != nil {
+		crSuppressedReasons = append(crSuppressedReasons, cr.Spec.ForProvider.SuppressionOptions.SuppressedReasons...)
```
`cr` should not be changed during `isUpTodate`.

kelvinwijaya replied:
have fixed this issue

kelvinwijaya commented on pkg/controller/sesv2/configurationset/setup.go:
```diff
@@ -0,0 +1,324 @@
+func isUpToDateDeliveryOptions(cr *svcapitypes.ConfigurationSet, resp *svcsdk.GetConfigurationSetOutput) bool {
+	// User may remove DeliveryOptions spec and expect an update
+	if cr.Spec.ForProvider.DeliveryOptions == nil && resp.DeliveryOptions != nil {
+		cr.Spec.ForProvider.DeliveryOptions = &svcapitypes.DeliveryOptions{}
```
have fixed this issue

MisterMX commented on pkg/controller/sesv2/configurationset/setup.go:
```diff
@@ -0,0 +1,305 @@
+			TlsPolicy:            cr.Spec.ForProvider.DeliveryOptions.TLSPolicy,
+		}
+		if _, err := e.client.PutConfigurationSetDeliveryOptionsWithContext(ctx, deliveryOptionsInput); err != nil {
+			return managed.ExternalUpdate{}, errors.Wrap(err, "update failed")
```
There should be a different error message for each `Wrap` to help identify where the error came from. `"update failed"` is too generic here.

MisterMX commented on pkg/controller/sesv2/configurationset/setup.go:
```diff
@@ -0,0 +1,305 @@
+			ReputationMetricsEnabled: cr.Spec.ForProvider.ReputationOptions.ReputationMetricsEnabled,
+		}
+		if _, err := e.client.PutConfigurationSetReputationOptionsWithContext(ctx, reputationOptionsInput); err != nil {
+			return managed.ExternalUpdate{}, errors.Wrap(err, "update failed")
```
There should be a different error message for each `Wrap` to help identify where the error came from. `"update failed"` is too generic here.

MisterMX commented on pkg/controller/sesv2/configurationset/setup.go:
```diff
@@ -0,0 +1,305 @@
+			SuppressedReasons:    suppresssedReasons,
+		}
+		if _, err := e.client.PutConfigurationSetSuppressionOptionsWithContext(ctx, supressOptionsInput); err != nil {
+			return managed.ExternalUpdate{}, errors.Wrap(err, "update failed")
```
There should be a different error message for each `Wrap` to help identify where the error came from. `"update failed"` is too generic here.

MisterMX commented on pkg/controller/sesv2/configurationset/setup.go:
```diff
@@ -0,0 +1,305 @@
+			CustomRedirectDomain: cr.Spec.ForProvider.TrackingOptions.CustomRedirectDomain,
+		}
+		if _, err := e.client.PutConfigurationSetTrackingOptionsWithContext(ctx, trackingOptionInput); err != nil {
+			return managed.ExternalUpdate{}, errors.Wrap(err, "update failed")
```
There should be a different error message for each `Wrap` to help identify where the error came from. `"update failed"` is too generic here.

kelvinwijaya replied:
resolved

MisterMX commented on pkg/controller/sesv2/configurationset/setup.go:
```diff
@@ -0,0 +1,305 @@
+			SendingEnabled:       cr.Spec.ForProvider.SendingOptions.SendingEnabled,
+		}
+		if _, err := e.client.PutConfigurationSetSendingOptionsWithContext(ctx, sendingOptionInput); err != nil {
+			return managed.ExternalUpdate{}, errors.Wrap(err, "update failed")
```
There should be a different error message for each `Wrap` to help identify where the error came from. `"update failed"` is too generic here.

kelvinwijaya replied:
resolved

MisterMX commented on pkg/controller/sesv2/emailidentity/setup.go:
```diff
@@ -0,0 +1,261 @@
+		EmailIdentity:        cr.Spec.ForProvider.EmailIdentity,
+	}
+	if _, err := e.client.PutEmailIdentityConfigurationSetAttributesWithContext(ctx, configurationSetAttributesInput); err != nil {
+		return managed.ExternalUpdate{}, errors.Wrap(err, "update failed")
```
There should be a different error message for each `Wrap` to help identify where the error came from. `"update failed"` is too generic here.

kelvinwijaya replied:
resolved

MisterMX commented on pkg/controller/sesv2/emailidentity/setup.go:
```diff
@@ -0,0 +1,261 @@
+		}
+
+		if _, err := e.client.PutEmailIdentityMailFromAttributesWithContext(ctx, mailFromAttributesInput); err != nil {
+			return managed.ExternalUpdate{}, errors.Wrap(err, "update failed")
```
There should be a different error message for each `Wrap` to help identify where the error came from. `"update failed"` is too generic here.

kelvinwijaya replied:
resolved

MisterMX commented on pkg/controller/sesv2/emailidentity/setup.go:
```diff
@@ -0,0 +1,261 @@
+			SigningAttributesOrigin: awsclient.String(dkimSigningAttributeExternal),
+		}
+		if _, err := e.client.PutEmailIdentityDkimSigningAttributesWithContext(ctx, dkimSigningAttributesInput); err != nil {
+			return managed.ExternalUpdate{}, errors.Wrap(err, "update failed")
```
There should be a different error message for each `Wrap` to help identify where the error came from. `"update failed"` is too generic here.

kelvinwijaya replied:
resolved

MisterMX commented on pkg/controller/sesv2/emailidentity/setup.go:
```diff
@@ -0,0 +1,261 @@
+			},
+		}
+		if _, err := e.client.PutEmailIdentityDkimSigningAttributesWithContext(ctx, dkimSigningAttributesInput); err != nil {
+			return managed.ExternalUpdate{}, errors.Wrap(err, "update failed")
```
There should be a different error message for each `Wrap` to help identify where the error came from. `"update failed"` is too generic here.

kelvinwijaya replied:
resolved

MisterMX reviewed (changes requested):
Looks almost good to me. Only some minor issues remaining. @guilhermef after solving these can you squash your changes into a single commit?

kelvinwijaya commented:
@MisterMX , thanks for your kind review and i have squashed the commit messages

MisterMX reviewed (approved):
@kelvinwijaya thank you very much for your contribution! LGTM.
